---
title: "Project Analyses"
author: "Kevin O'Connor"
date: "3/25/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Clear console
rm(list=ls())
stringsAsFactors <- FALSE

# Libraries
library(tidyverse)
library(survival)
library(circlize)
library(DESeq2)
library(keras)
library(preprocessCore)
library(R.utils)

# Directories
homedir <- "/Users/kevinoconnor/Documents/School/CancerGenomics/Project"
codedir <- file.path(homedir, "UNC_CancerGenomics_Spring19_Group2")
datadir <- file.path(homedir, "Data")
dcmdir  <- "/Users/kevinoconnor/Documents/Research/DCM/Differential-Correlation-Mining/DCM/R"
sourceDirectory(dcmdir, modifiedOnly = FALSE)

# Read data.
source(file.path(codedir, "DataLoader.R"))
data_list <- DataLoader(datadir)
rnaseq <- data_list$rnaseq
clinical   <- data_list$clinical
```

```{r check_qn}
# Checking for upper quartile normalization.
apply(tcga_rnaseq, 2, function(x) quantile(x, probs = 0.5))[1:20]
apply(target_rnaseq, 2, function(x) quantile(x, probs = 0.5))[1:20]
```

```{r data_summary}
library(vioplot)
# Age plot.
png(file.path(homedir, "age_plot.png"))
plot(density(clinical$age), main="Age in Days", xlab="Age in Days")
dev.off()
# Survival time plot.
png(file.path(homedir, "survival_time_plot.png"))
plot(density(na.omit(clinical$survtime)), main="Survival Time in Days", xlab="Survival Time in Days")
dev.off()
# Aneuploidy score
png(file.path(homedir, "aneuploidy_plot.png"))
plot(density(na.omit(clinical$Aneuploidy.Score)), main="Aneuploidy Score", xlab="Aneuploidy Score")
dev.off()
# Violin age plot.
png(file.path(homedir, "age_violin_plot.png"))
vioplot(clinical$age[which(substr(clinical$Row.names, start=1, stop=6) == "TARGET")],
        clinical$age[-which(substr(clinical$Row.names, start=1, stop=6) == "TARGET")], 
        main="Age in Days", names=c("TARGET", "TCGA"), col="lightblue", ylab="Age in Days")
dev.off()
# Histograms of VEGF
png(file.path(homedir, "vegfa_plot.png"))
hist(as.numeric(rnaseq["VEGFA", ]), xlim=c(-4, 4), main="VEGFA Expression", xlab="Expression")
dev.off()
png(file.path(homedir, "vegfb_plot.png"))
hist(as.numeric(rnaseq["VEGFB", ]), xlim=c(-4, 4), main="VEGFB Expression", xlab="Expression")
dev.off()
png(file.path(homedir, "vegfc_plot.png"))
hist(as.numeric(rnaseq["VEGFC", ]), xlim=c(-4, 4), main="VEGFC Expression", xlab="Expression")
dev.off()
```


```{r dcm_gender}
gender <- clinical$gender[match(colnames(rnaseq), clinical$Row.names)] %>% as.vector()
# Run DCM
dcm_male_v_female <- DCM(rnaseq[, which(gender == "male")], rnaseq[, which(gender == "female")], max.groups=2, echo=TRUE)
dcm_female_v_male <- DCM(rnaseq[, which(gender == "female")], rnaseq[, which(gender == "male")], max.groups=1, echo=TRUE)
dcm_male_v_female_sets <- lapply(dcm_male_v_female$DC_sets, function(x) sort(rownames(rnaseq)[x]))
dcm_female_v_male_sets <- lapply(dcm_female_v_male$DC_sets, function(x) sort(rownames(rnaseq)[x]))
# GO analysis
go_analysis(dcm_male_v_female, rownames(female_rnaseq))
go_analysis(dcm_female_v_male_set, rownames(male_rnaseq))
```

```{r dcm_race}
# Prepare data
race <- clinical$race[match(colnames(rnaseq), clinical$Row.names)] %>% as.vector()
white_inds <- which(race == "white")
black_inds <- which(race == "black or african american")
# Run DCM
#dcm_white_v_afram <- DCM(rnaseq[, white_inds], rnaseq[, black_inds], max.groups=1, echo=TRUE)
#dcm_afram_v_white <- DCM(rnaseq[, black_inds], rnaseq[, white_inds], max.groups=1, echo=TRUE)
#dcm_afram_v_white_set <- rownames(white_rnaseq)[dcm_afram_v_white$DC_sets[[1]]]
# GO analysis.
#go_analysis(dcm_afram_v_white_set, rownames(white_rnaseq))
```

```{r dcm_survtime}
# Prepare data
time_lower_cut <- 365
time_upper_cut <- 5*365
survtime <- clinical$survtime[match(colnames(rnaseq), clinical$Row.names)] %>% as.vector()
short_inds <- which(survtime <= time_lower_cut)
long_inds <- which(survtime > time_lower_cut & survtime <= time_upper_cut)
# Run DCM
#dcm_long_v_short <- DCM(rnaseq[, short_inds], rnaseq[, long_inds], max.groups=1, echo=TRUE)
dcm_short_v_long <- DCM(rnaseq[, long_inds], rnaseq[, short_inds], max.groups=1, echo=TRUE)
#dcm_long_v_short_set <- rownames(rnaseq)[dcm_long_v_short$DC_sets[[1]]]
dcm_short_v_long_sets <- lapply(dcm_short_v_long$DC_sets, function(x) sort(rownames(rnaseq)[x]))
# Go analysis
#go_analysis(dcm_long_v_short_set, rownames(rnaseq))
go_analysis(dcm_short_v_long_set[[1]], rownames(rnaseq))
```

```{r dcm_aneuploidy}
# Prepare data
an_score <- clinical$Aneuploidy.Score[match(colnames(rnaseq), clinical$Row.names)] %>% as.vector()
low_inds <- which(an_score == 0)
high_inds <- which(an_score > 0)
# Run DCM
#dcm_lowas_v_highas <- DCM(rnaseq[, low_inds], rnaseq[, high_inds], max.groups=1, echo=TRUE)
#dcm_highas_v_lowas <- DCM(rnaseq[, high_inds], rnaseq[, low_inds], max.groups=1, echo=TRUE)
#dcm_short_v_long_sets <- lapply(dcm_short_v_long$DC_sets, function(x) sort(rownames(rnaseq)[x]))
# Go analysis
#go_analysis(dcm_long_v_short_set, rownames(rnaseq))
#go_analysis(dcm_short_v_long_set[[1]], rownames(rnaseq))
```




```{r stuff}
fileConn <- file(file.path(homedir, "dcm_sets.txt"), open="wb")
writeLines("Male vs. Female", fileConn)
for(s in dcm_male_v_female_sets){
  writeLines(sort(s), fileConn)
  writeLines("", fileConn)  
}
writeLines("Female vs. Male", fileConn)
for(s in dcm_female_v_male_sets){
  writeLines(sort(s), fileConn)
  writeLines("", fileConn)  
}
writeLines("Short vs. Long Survival Time", fileConn)
for(s in dcm_short_v_long_sets){
  writeLines(sort(s), fileConn)
  writeLines("", fileConn)
}
close(fileConn)
```



```{r compare_to_vascular_signatures}
vasc_sign <- read.csv(file.path(homedir, "vascular_signatures.csv"), header=TRUE)
vasc_genes <- as.vector(vasc_sign$Gene.symbol)
# Male vs Female
print("Male vs Female")
for(s in dcm_male_v_female_sets){
  print(intersect(s, vasc_sign$Gene.symbol))
}
# Female vs Male
print("Female vs Male")
for(s in dcm_female_v_male_sets){
  print(intersect(s, vasc_sign$Gene.symbol))
}
# Short vs long survival time
print("Short vs Long Survival Time")
for(s in dcm_short_v_long_sets){
  print(intersect(s, vasc_sign$Gene.symbol))
}

```
